0-module-and-module-source.emu

```diff
@@ -0,0 +1,661 @@
+              1. Let _importEntries_ be _sourceModule_.[[ImportEntries]].
+              1. Let _localExportEntries_ be _sourceModule_.[[LocalExportEntries]].
+              1. Let _indirectExportEntries_ be _sourceModule_.[[IndirectExportEntries]].
+              1. Let _starExportEntries_ be _sourceModule_.[[StarExportEntries]].
+              1. Let _hostDefined_ be _sourceModule_.[[HostDefined]].
+              1. Let _realm_ be the current Realm Record.
+              1. Return Source Text Module Record { [[Realm]]: _realm_, [[Environment]]: ~empty~, [[Namespace]]: ~empty~, [[CycleRoot]]: ~empty~, [[HasTLA]]: _async_, [[AsyncEvaluation]]: *false*, [[TopLevelCapability]]: ~empty~, [[AsyncParentModules]]: &laquo; &raquo;, [[PendingAsyncDependencies]]: ~empty~, [[Status]]: ~unlinked~, [[EvaluationError]]: ~empty~, [[HostDefined]]: _hostDefined_, [[ECMAScriptCode]]: _body_, [[Context]]: ~empty~, [[ImportMeta]]: ~empty~, [[RequestedModules]]: _requestedModules_, [[ImportEntries]]: _importEntries_, [[LocalExportEntries]]: _localExportEntries_, [[IndirectExportEntries]]: _indirectExportEntries_, [[StarExportEntries]]: _starExportEntries_, [[DFSIndex]]: ~empty~, [[DFSAncestorIndex]]: ~empty~, [[ModuleInstance]]: *undefined* }.
+            </emu-alg>
+        </emu-clause>
+
+        <emu-clause id="sec-resolvemodulerecorddependency" type="abstract operation">
+            <h1>
+                ResolveModuleRecordDependency (
+                    _moduleRecord_: a Module Record,
+                    _specifier_: a |ModuleSpecifier| String,
+                    _promiseCapability_: a PromiseCapability Record,
+                ): ~unused~
+            </h1>
+            <dl class="header">
+              <dt>description</dt>
+              <dd>It provides the concrete Module Record subclass instance that corresponds to _specifier_ occurring within the module represented by _moduleRecord_.</dd>
+            </dl>
+            <emu-alg>
+              1. Assert: _moduleRecord_ is a Module Record.
+              1. Assert: _moduleRecord_.[[ModuleInstance]] is an Object.
+              1. Assert: ResolveModuleRecordDependency has not been invoked with _moduleRecord_ and _specifier_ pair.
+              1. Let _moduleInstance_ be _moduleRecord_.[[ModuleInstance]].
+              1. Let _importHook_ be _moduleInstance_.[[ImportHook]].
+              1. Let _completion_ be Completion(Call(_importHook_, *undefined*, « _specifier_ »).
+              1. IfAbruptRejectPromise(_completion_, _promiseCapability_).
+              1. Let _importHookPromise_ be ? PromiseResolve(%Promise%, _completion_.[[Value]]).
+              1. Let _fulfilledClosure_ be a new Abstract Closure with parameters (_result_) that captures _moduleRecord_ and _promiseCapability_ and performs the following steps when called
+                1. If _result_ is an Object that has a [[ImportHook]] internal slot, then
+                  1. Let _moduleRecord_ be _result_.[[Module]].
```

Comment:
@nicolo-ribaudo @Jack-Works, I have two issues here:

1. if the result of calling `[[ImportHook]]` function must always be a Module Instance, then how can we resolve to `fs` or some other module instance that delegates the resolution to the browser? That seems like a problem. I don't know if the issue is here, or somewhere else, but definitely needs to be addressed.
2. where do we go into the recursive resolution? in `HostImportModuleRecordDynamically` we only talk about the "unique [ModuleSpecifier](https://tc39.es/ecma262/#prod-ModuleSpecifier) associated to the moduleRecord", never about the transitive dependencies. any suggestion?

Reply:
One option here might be to support having import hook return a `ModuleSource` which is then instanced against its primary default instance for the host context? Would need to think through the edge cases though....

Recursive resolution happens through `HostResolveImportedModule` calls in `InnerModuleLinking` and `InnerModuleEvaluation`. I really liked @nicolo-ribaudo's approach in https://nicolo-ribaudo.github.io/modules-import-hooks-refactor/ of having this being a sync lookup against the resolution list on the module record itself though to more clearly separate the sync and async iterations.

Reply:
1. My hope is that import reflection could also give `Module`s other than `ModuleSource`s <small>(the syntax is just an idea)</small>:
   ```js
   import module jsM from "x.js";
   import source jsS from "x.js";
   import module wasmM from "y.wasm";
   import source wasmS from "y.wasm";

   jsM instanceof Module;
   jsS instanceof ModuleSource;
   wasmM instanceof Module;
   wasmS instanceof WebAssembly.Module; // WebAssembly.Module respects the "module source interface", whatever that means

   jsM.source === jsS;
   wasmM.source === wasmS;
    ```

   Then, an import hook might be implemented like this:
   ```js
   function importHook(specifier, importMeta) {
     if (specifier === "fs") return import("fs", { reflection: "module" });
     
     const source = // get the module source
     const innerImportMeta = // get the import.meta
     return new Module(source, importHook, innerImportMeta);
   }
   ```
2. As @guybedford said, the current ECMA-262, we explicitly handle the recursive resolution: we hide the fact that "real resolution" can be asynchronous, and we pretend that the resolution happens in `HostResolveImportedModule`. We call it recursively for every transitive dependency in `InnerModuleLinking`, which is triggered by an (implicit) call to `.Link()` of the top-level imported module in `HostImportModuleDynamically`.
   We probably can keep hand waving in this proposal spec (we need a "complete spec text" only when asking for stage 3, and this PR is already very good for asking for stage 2), but a proper solution will look a lot like https://nicolo-ribaudo.github.io/modules-import-hooks-refactor/#sec-LoadRequestedModules (or, I push for merging https://nicolo-ribaudo.github.io/modules-import-hooks-refactor without waiting for a proposal, and then you just edit that in this proposal to call `[[ImportHook]]` instead of `HostLoadImportedModule`)

Reply:
@nicolo-ribaudo on (1), this is an interesting proposal although does introduce more complexity. That said it very much resolves some current concerns through the separation. Would the module instance reflection be unexecuted if it hasn't already been executed? But would the dependencies be fully preloaded? If so, that may well make it suitable for deferred evaluation mechanisms....

Furthermore, passing an instance to another realm in this way with linkage might enable an easy solution to resolution sharing without having to re-instrument a loader on the other realm...

Would be interested to hear more details if it might be able to address the above.